tests/commands/test_store_client.py

```diff
@@ -434,3 +438,116 @@ def test_client_errorparsing_bad_structure():
+
+
+def test_client_push_simple_ok(caplog, tmp_path, capsys):
+    """Happy path for pushing bytes."""
+    caplog.set_level(logging.DEBUG, logger="charmcraft.commands")
+
+    # fake some bytes to push
+    test_filepath = tmp_path / 'supercharm.bin'
+    with test_filepath.open('wb') as fh:
+        fh.write(b"abcdefgh")
+
+    def fake_pusher(monitor):
+        """Push bytes in sequence, doing verifications in the middle."""
+        total_to_push = monitor.len  # not only the saved bytes, but also headers and stuff
+
+        # one batch
+        monitor.read(20)
+        captured = capsys.readouterr()
+        assert captured.out == "Uploading... {:.2f}%\r".format(100 * 20 / total_to_push)
+
+        # another batch
+        monitor.read(20)
+        captured = capsys.readouterr()
+        assert captured.out == "Uploading... {:.2f}%\r".format(100 * 40 / total_to_push)
+
+        # check monitor is properly built
+        assert isinstance(monitor.encoder, MultipartEncoder)
+        srctype, fh, ctype = monitor.encoder.fields['binary']
+        assert srctype == 'filename'
+        assert fh.name == str(test_filepath)
+        assert ctype == "application/octet-stream"
+
+        content = json.dumps(dict(successful=True, upload_id='test-upload-id'))
+        return FakeResponse(content=content, status_code=200)
+
+    with patch('charmcraft.commands.store.client._storage_push', fake_pusher):
+        Client().push(test_filepath)
+
+    # check proper logs
+    expected = [
+        "Starting to push {}".format(str(test_filepath)),
+        "Uploading bytes ended, id test-upload-id",
+    ]
+    assert expected == [rec.message for rec in caplog.records]
+
+
+def test_client_push_response_not_ok(tmp_path):
+    """Didn't get a 200 from the Storage."""
+    # fake some bytes to push
+    test_filepath = tmp_path / 'supercharm.bin'
+    with test_filepath.open('wb') as fh:
+        fh.write(b"abcdefgh")
+
+    with patch('charmcraft.commands.store.client._storage_push') as mock:
+        mock.return_value = FakeResponse(content='had a problem', status_code=500)
+        with pytest.raises(CommandError) as cm:
+            Client().push(test_filepath)
+        assert str(cm.value) == "Failure when pushing file: [500] 'had a problem'"
+
+
+def test_client_push_response_unsuccessful(tmp_path):
+    """Didn't get a 200 from the Storage."""
+    # fake some bytes to push
+    test_filepath = tmp_path / 'supercharm.bin'
+    with test_filepath.open('wb') as fh:
+        fh.write(b"abcdefgh")
+
+    with patch('charmcraft.commands.store.client._storage_push') as mock:
+        raw_content = dict(successful=False, upload_id=None)
+        mock.return_value = FakeResponse(content=json.dumps(raw_content), status_code=200)
+        with pytest.raises(CommandError) as cm:
+            Client().push(test_filepath)
+        assert str(cm.value) == "Server error when pushing file: " + repr(raw_content)
```

Comment:
Server error while sending file? while pushing file?

Reply:
Changed this and other errors to use "while", thanks. I like "pushing" as it represents the used method.